drivers/stepper/adi/tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,608 @@
+	struct spi_dt_spec spi;
+	const uint32_t clock_frequency;
+#if CONFIG_STEPPER_ADI_TMC5041_INT
+	struct gpio_dt_spec int_pin_spec;
+	/* connected motors */
+	const struct device **motors;
+	/* number of connected motors */
+	const uint8_t motor_count;
+#endif
+};
+
+struct tmc5041_stepper_data {
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	struct k_work_delayable stallguard_dwork;
+	/* device pointer required to access config in k_work */
+	const struct device *stepper;
+#endif
+#ifdef CONFIG_STEPPER_ADI_TMC5041_INT
+	struct k_poll_signal *async_signal;
+#endif
+};
+
+struct tmc5041_stepper_config {
+	const uint8_t index;
+	const uint16_t default_micro_step_res;
+	/* parent controller required for bus communication */
+	const struct device *controller;
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	int8_t stallguard_threshold;
+#endif
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+	const struct tmc_ramp_generator_data default_ramp_config;
+#endif
+};
+
+static int tmc5041_write(const struct device *dev, const uint8_t reg_addr, const uint32_t reg_val)
+{
+	const struct tmc5041_config *config = dev->config;
+	struct tmc5041_data *data = dev->data;
+	const struct spi_dt_spec bus = config->spi;
+	int status = k_mutex_lock(&data->mutex, K_MSEC(100));
+
+	if (!status) {
+		status = tmc_spi_write_register(&bus, reg_addr, reg_val);
+		k_mutex_unlock(&data->mutex);
+	}
+	return status;
+}
+
+static int tmc5041_read(const struct device *dev, const uint8_t reg_addr, uint32_t *reg_val)
+{
+	const struct tmc5041_config *config = dev->config;
+	struct tmc5041_data *data = dev->data;
+	const struct spi_dt_spec bus = config->spi;
+	int status = k_mutex_lock(&data->mutex, K_MSEC(100));
+
+	if (!status) {
+		status = tmc_spi_read_register(&bus, reg_addr, reg_val);
+		(void)k_mutex_unlock(&data->mutex);
+	}
+	return status;
+}
+
+static void calculate_velocity_from_hz_to_fclk(const struct device *dev, const uint32_t velocity_hz,
+					       uint32_t *const velocity_fclk)
+{
+	const struct tmc5041_config *config = dev->config;
+
+	if (config->clock_frequency == 0) {
+		LOG_ERR("Clock frequency cannot be zero");
+		return;
+	}
+
+	*velocity_fclk =
+		((uint64_t)(velocity_hz) << TMC5041_CLOCK_FREQ_SHIFT) / config->clock_frequency;
+	LOG_DBG("Stepper motor controller %s velocity: %d Hz, velocity_fclk: %d", dev->name,
+		velocity_hz, *velocity_fclk);
+}
+
+static int set_async_signal(const struct device *dev, struct k_poll_signal *async)
+{
+	struct tmc5041_stepper_data *data = dev->data;
+
+	if (async) {
+#ifdef CONFIG_STEPPER_ADI_TMC5041_INT
+		if (data->async_signal) {
+			k_poll_signal_reset(data->async_signal);
+		}
+		data->async_signal = async;
+#else
+		ARG_UNUSED(data);
+		LOG_WRN("STEPPER_ADI_TMC5041_INT not enabled, async signal not supported");
+#endif
+	}
+	return 0;
+}
+
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+static int32_t print_stall_status(const struct device *dev)
+{
+	const struct tmc5041_stepper_config *config = dev->config;
+
+	int32_t sg_result = 0;
```

Comment:
no need to init this, or rather it should not be needed, right now there's some situation where an the read fails and this returns zero with no logging or anything